extension-manifest-v3/src/background/reporting.js

```diff
@@ -216,26 +216,32 @@ const reporting = new Reporting({
   communication,
 });
 
-observe('terms', (terms) => {
-  if (terms) {
-    reporting.init().catch((e) => {
-      console.warn(
-        'Failed to initialize reporting. Leaving the module disabled and continue.',
-        e,
-      );
-    });
-  } else {
-    reporting.unload();
-  }
-});
+let pendingStartup = (async function () {
+  await observe('terms', async (terms) => {
+    try {
+      if (terms) {
+        try {
+          await reporting.init();
+        } catch (e) {
+          console.warn(
+            'Failed to initialize reporting. Leaving the module disabled and continue.',
+            e,
+          );
+        }
+      } else {
+        reporting.unload();
+      }
+    } finally {
+      pendingStartup = null;
+    }
+  });
+})();
```

Comment:
You can simply pass here `pendingStartup = null`, as you `await observe()..`, and remove `try {} finally {}` wrapper.

See here: https://github.com/ghostery/ghostery-extension/blob/main/extension-manifest-v3/src/background/adblocker.js#L51

Reply:
Now I think, it might be a good idea to create a global util function, which takes an async function, and returns a promise, which resolves, or rejects in some threshold. Then in 277 it could just safely await the promise.

Something like:

```js
const setup = asyncSetup(async () => { ... });

...
if (setup.promise) {
  await setup.promise;
}
```

Obviously, the util should return an object with promise, which is cleared.

Reply:
I think that would break: "pendingStartup" resolves immediately while the callback is still pending. For instance:

```
let pendingStartup = (async function () {
  await observe('terms', async (terms) => {
    await delay(10000000);  // <--- simulate that reporting hangs 
    if (terms) {
      try {
        await reporting.init();
      } catch (e) {
        console.warn(
          'Failed to initialize reporting. Leaving the module disabled and continue.',
          e,
        );
      }
    } else {
      reporting.unload();
    }
  });
  pendingStartup = null;
})();
````

That was why I moved the "pendingStartup = null" at the end of the callback execution. Or did I misunderstood what you meant?

Reply:
I debugged into (with changed version above). It hits another bug is related to the change-detection in line https://github.com/ghostery/ghostery-extension/blob/b05fa7de45e02ad52f86b45f025f2328348f53fc/extension-manifest-v3/src/store/options.js#L159

In a nutshell, adding it to the observer and later calling the wrapper races.

The wrapper function is called twice:
1. As part of the observers set:https://github.com/ghostery/ghostery-extension/blob/b05fa7de45e02ad52f86b45f025f2328348f53fc/extension-manifest-v3/src/store/options.js#L92
2. When it is now called from reporting: https://github.com/ghostery/ghostery-extension/blob/main/extension-manifest-v3/src/store/options.js#L178, the wrapper has been already execute (part of the observer set)

Only the first call blocks (but in the observer trigger, it's ignored: https://github.com/ghostery/ghostery-extension/blob/b05fa7de45e02ad52f86b45f025f2328348f53fc/extension-manifest-v3/src/store/options.js#L92). When it now runs for the second time, the value matches the old value and the call is skipped (value=false and options[property]=false): https://github.com/ghostery/ghostery-extension/blob/b05fa7de45e02ad52f86b45f025f2328348f53fc/extension-manifest-v3/src/store/options.js#L159

The effect is that the promise now immediately resolves ("observe" resolved), even though the function is still executing.

Reply:
It's the question how to proceed. If we fix the race in "observe", I also think it could be changed to the alternative version. Since from the caller perspective, the semantic that the first execution is guaranteed to have finished is more intuitive, I would prefer that.

But maybe it should be fixed in its own PR, since it has global effect. And once it's fixed, the try-finally construction with de-facto serves as a workaround could be removed.

Addressing the race is a bit subtle. For instance, postponing the registration of the observer could fix it in this case, but may break other code that currently relies on that behavior.

Reply:
Calling the `observe()` function, and then `fn` by internal options implementation is not related. It works correctly in that it won't call the `fn` again if the value hasn't changed.

It's a simple thing, `observe()` returns a promise, which guarantees that it is resolved after calling `fn` (at least once). This is the only thing that matters here. If the observed value changes later/in the middle/whatever time, it will call `fn` again with a defined side effect. 

The core "problem" of the implementation is that some features, which can be required at any time require configuration, which is getting async (`chrome.storage.local`). 

The observe method rejects if options cannot be fetched, or when `fn` rejects. I am not even sure if there is a need for `Promise.race` with a timeout, as there is no option to hang that promise.

In my opinion the pattern in `adblocker.js` is fine, but as we have more places, which need that structure, it might be a good idea to create a helper util, which covers setting promise, and then clear it out (this is also an optimization, that events don't have to wait every time for already fulfilled promise).

Reply:
My point is that with the current implementation, observe is not guaranteed to block (see above). Only on the first call, but there are multiple callers out of your control, so you can't rely on it.

The PR in the current state is not affected, since it waits in the call. But if I would make the change that you requested (replacing try-finally), it would hit the bug.

Reply:
To make it clear what I mean, this other PR fixes the problem that I tried to describe: https://github.com/ghostery/ghostery-extension/pull/1129

The other PR would guarantee a caller that the wrapper ran once after "await observe" resolves. I tested with that fix, and then I can remove the try-finally constructor and move the nulling out of the function.

Reply:
Yes, you have a guarantee that the passed function to the `observe()` has finished - https://github.com/ghostery/ghostery-extension/blob/main/extension-manifest-v3/src/store/options.js#L178, when you initialize observer by calling `await observe(fn)`.

Reply:
The only thing that might be changed is to move that line after the initial call - https://github.com/ghostery/ghostery-extension/blob/main/extension-manifest-v3/src/store/options.js#L171, just before the `return` statement. It would avoid calling fn while there is an initial call (which is pointless).

Reply:
> Yes, you have a guarantee that the passed function to the `observe()` has finished - https://github.com/ghostery/ghostery-extension/blob/main/extension-manifest-v3/src/store/options.js#L178, when you initialize observer by calling `await observe(fn)`.

That is not correct. It needs a fix that like one here: https://github.com/ghostery/ghostery-extension/pull/1129

At the moment, we don't have the guarantee. The reason is the race though the observer set. It can be reproduce with the change above (https://github.com/ghostery/ghostery-extension/pull/1115#discussion_r1172593303) that simulates a delay.

Reply:
Included the fix now for the observer race (by moving the registration), and applied the simplification now (getting rid of try-finally).